### PR TITLE
Condition Sharing

### PR DESCRIPTION
**Pull request by chris-pardy**
This introduces the concept of Condition sharing which allows regular conditions that are given a well-known name and to be added to the engine. Rules can reference the condition by including a `condition` property on a condition.

example based on #296 
```js
engine.setCondition('isMinor', {
  all: [
    {
        fact: 'age',
        operator: 'lessThan',
       value: { fact: 'drinkingAge' }
    }
  ]
});

engine.addRule({
   conditions: {
      all: [
          { not: { condition: 'isMinor' } },
         {
             fact: 'currentTime',
             operator: 'lessThanInclusive',
             value: { fact: 'lastCall' }
         }
     ]
  },
    event: {
      type: 'AlcoholicBeverageSelection'
   }
});

// condition references can be top-level conditions so it's possible for a rule to just be a condition reference.
engine.addRule({
 conditions: {
    condition: 'isMinor'
  },
  event: {
     type: 'SoftDrinkSelection'
  }
})
```

This addresses #336, #296, and #263

A few things related to this:

### Why do we need this at-all
The ability to capture common conditions and compose rules is useful as a first-class function of this rules engine. As rules reach increased levels of complexity common patterns emerge that should be re-used rather than repeated. One option to do this is to add a "pre-processing" step outside the rules engine that would build up rules from component parts. However this means that someone inspecting the rules as they're run would need to understand the behavior of this pre-processing. Additionally having this be supported by the rules engine allows for more universal editing tools to adopt this as something that can be edited. 

### Condition references vs. rule-chaining
The primary difference here is that the composition can be communicated using JSON because conditions are fully expressible as JSON. This prevents the javascript code from needing to know about specific rules to setup specific event handlers. This allows us to cleanly separate javascript code that configures the rules engine from the rules themselves. Additionally there is no need to specify rule priority when using conditions as the condition references will be resolved at runtime.

### Why condition references and not something with facts
I had originally suggested using a "conditional fact" or other construct to capture the idea of re-used logic. This was nice because it matched concepts that already existed in the rules engine. However it fails the equivalency test. Extracting re-used conditions into a fact would cause the rule results to be different. I liked that the use of Conditions and the approach of de-referencing the condition references at runtime would keep the same structure to the rule results even after refactoring.

### Concern with duplicated work.
Condition references are dereferenced and evaluated within each rule they're part of. That means that if a condition is used in 10 rules it will be evaluated 10 times. I'm not concerned about this as the majority of the work will be captured by the Fact value caching in the almanac. The work that is left to be done ought to be very quick. While the possibility of result caching does exist as a future improvement adding it here seemed unnecessarily complex.

### CODE REVIEW

File: docs/rules.md [hunk 1]
```diff
@@ -174,7 +175,30 @@ let rule = new Rule({
 })
 ```
 
-Notice in the second example how `all`, `any`, and 'not' can be nested within one another to produce complex boolean expressions.  See the [nested-boolean-logic](../examples/02-nested-boolean-logic.js) example.
+Notice in the second example how `all`, `any`, and `not` can be nested within one another to produce complex boolean expressions.  See the [nested-boolean-logic](../examples/02-nested-boolean-logic.js) example.
+
+### Shared Conditions
+
+Rules my reference shared conditions based on their name.
```
Comment (RoboCafaz) [inline docs/rules.md:182]: Rules "may" reference...

File: docs/rules.md [hunk 2]
```diff
@@ -136,7 +137,7 @@ See the [hello-world](../examples/01-hello-world.js) example.
 
 ### Boolean expressions: `all`, `any`, and `not`
 
-Each rule's conditions *must* have an `all` or `any` operator containing an array of conditions at its root or a `not` operator containing a single condition.  The `all` operator specifies that all conditions contained within must be truthy for the rule to be considered a `success`.  The `any` operator only requires one condition to be truthy for the rule to succeed. The `not` operator will negate whatever condition it contains.
+Each rule's conditions *must* reference a shared condition or have an `all` or `any` operator containing an array of conditions at its root or a `not` operator containing a single condition.  The `all` operator specifies that all conditions contained within must be truthy for the rule to be considered a `success`.  The `any` operator only requires one condition to be truthy for the rule to succeed. The `not` operator will negate whatever condition it contains.
```
Comment (RoboCafaz) [inline docs/rules.md:140]: Nit: this section is about the boolean operators but the first condition it references is the shared conditions. Consider rewording.

**Review by RoboCafaz — APPROVED** [review]
A very simple solution for adding additional composability options to the rules engine. I'm a fan.

The lack of result caching is a bit of a bummer but that's something that could be iterated upon in future updates.

File: docs/rules.md [hunk 3]
```diff
@@ -174,7 +175,30 @@ let rule = new Rule({
...
+})
+```
+
+Before running the rule the shared condition should be added to the engine.
+
+```js
+engine.addSharedCondition('sharedCondition', { /* conditions */ });
+```
+
+Shared conditions must start with `all`, `any`, `not`, or reference a shared condition.
```
Comment (matthew-mahony) [inline docs/rules.md:201]: nit: add a test or example of a shared condition referencing another shared condition

File: test/engine-shared-conditions.test.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
+'use strict'
+
+import sinon from 'sinon'
+import engineFactory from '../src/index'
+
+describe('Engine: shared conditions', () => {
```
Comment (matthew-mahony) [inline test/engine-shared-conditions.test.js:6]: Consider adding more test cases:
- Using multiple shared conditions
- Trying to use a shared condition that doesn't exist

File: src/engine.js [hunk 5]
```diff
@@ -92,6 +94,27 @@ class Engine extends EventEmitter {
...
 
+  /**
+   * Adds or updates a shared condition
+   * @param {string} name - the name of the shared condition to be referenced by rules.
+   * @param {object} conditions - the conditions to use when the shared condition is referenced.
+   */
+  addSharedCondition (name, conditions) {
+    if (!name) throw new Error('Engine: addSharedCondition() requires name')
+    if (!conditions) throw new Error('Engine: addSharedCondition() requires conditions')
+    this.sharedConditions.set(name, new Condition(conditions))
```
[the remark this answers is not in the document]
Comment (matthew-mahony) [reply src/engine.js:105]: Also consider adding log lines I see in other places, e.g. `debug(`engine:: addSharedCondition name:${name}`)`

**Review by matthew-mahony — APPROVED** [review]
Such a simple solution to something that will be super useful!

File: src/engine.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (chris-pardy) [reply src/engine.js:105]: I had originally created `add`, `update`, and `delete` but the `add` and `update` functions were effectively the same. I had also played with calling this function `upsertSharedCondtion` or `addOrUpdateSharedCondition` but they seemed too wordy and didn't fit with the `add` verb being use for Operators and Facts.

My initial take here is that I should just add documentation / jsdoc that makes it clear that `add` is an overwrite and leave the usecase for adding without overwriting to emerge naturally (if at all).

[the remark this answers is not in the document]
Comment (CacheControl) [reply src/engine.js:105]: It's a good call out. I think it's fine to allow this to overwrite. In general I was probably over-zealous in throwing too often in situations like this, and wish I'd opted for more of an upsert strategy that would have eliminated helpers like `updateRule()`.

**Review by CacheControl — COMMENTED** [review]
Underlying implementation looks good.

Per comments let's think hard about the naming and api we're exposing.

**Comment by CacheControl** [comment]
I think you'll see to add a line to the [toJSON](https://github.com/CacheControl/json-rules-engine/blob/master/src/condition.js#L42) helper in order to ensure the `uses: ` property serializes

File: src/engine.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (chris-pardy) [reply src/engine.js:105]: I thought about this and I'm going to go with `setSharedCondition` that more clearly implies the update or add behavior and it matches with the underlaying `Map` method.

File: docs/engine.md [hunk 6]
```diff
@@ -172,6 +174,71 @@ engine.addOperator('startsWithLetter', (factValue, jsonValue) => {
...
+      value: { fact: 'now' }
+    }
+  ]
+});
+
+engine.addRule({
+  condtions: {
+    all: [
+      {
+        uses: 'validLogin'
```
[the remark this answers is not in the document]
Comment (chris-pardy) [reply docs/engine.md:202]: Yeah the more time I spend with it the less I like `uses`. I'm fine with either `condition` or `ref` as I think there a more natural term. The one thing with `condition` is that it means a rule could be `{ conditions: { condition: <id> } }` how do we feel about that?

File: docs/engine.md [hunk 7]
```diff
@@ -172,6 +174,71 @@ engine.addOperator('startsWithLetter', (factValue, jsonValue) => {
 engine.removeOperator('startsWithLetter');
 ```
 
+### engine.addSharedCondition(String name, Object conditions)
+
+Adds a shared condition to the engine. Rules may include references to this shared condition.
+
+```javascript
+engine.addSharedCondition('validLogin', {
```
[the remark this answers is not in the document]
Comment (chris-pardy) [reply docs/engine.md:182]: I was doing some googling to try to see what this was called in other rules engines and there doesn't seem to be a good name for these in wide use. So I'm ok with `condition` as the name I actually like that in a future state this could be used to add new `Condition` subclass constructors which would let make it even more extensible.

**Comment by chris-pardy** [comment]
@CacheControl all comments addressed, looking for a final sign-off before I merge this and put together a point release.

**Review by CacheControl — APPROVED** [review]
